File: 백준/Silver/9461. 파도반 수열/파도반 수열.py
import sys
input = sys.stdin.readline

N = int(input())

maximum = 0
num = [0 for _ in range(N)]

for i in range(N) :
    num[i] = int(input())
    maximum = max(num[i], maximum+1)

result = [1, 1, 1, 2, 2]

for i in range(5, maximum+1) :
    result.append(result[i - 1] + result[i - 5])


for i in range(N) :
    print(result[num[i]-1])

File: 백준/Silver/9461. 파도반 수열/파도반 수열.java
import java.io.*;
import java.util.*;

public class Main {
	static int N;
	static int[] num;
	static long[] result;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		N = Integer.parseInt(st.nextToken());
		
		num = new int[N];
		
		int max = 5;
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			num[i] = Integer.parseInt(st.nextToken());
			max = Math.max(max, num[i]);
		}

		result = new long[max+1];
		result[0] = 1;
		result[1] = 1;
		result[2] = 1;
		result[3] = 2;
		result[4] = 2;
		
		for (int i = 5; i <= max; i++) {
			result[i] = (result[i-1] + result[i-5]);
		}
		
		for (int i = 0; i < N; i++) {
			sb.append(result[num[i] - 1] + "\n");
		}
		
		System.out.println(sb);
	}
}